python.d/ntpd.chart.py

```diff
@@ -0,0 +1,383 @@
+        +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+        |            Offset             |            Count              |
+        +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+        """
+        version = 2
+        sequence = 1
+        status = 0
+        offset = 0
+        count = 0
+        header = struct.pack(HEADER_FORMAT, (version << 3 | MODE), OPCODES[operation],
+                             sequence, status, associd, offset, count)
+        return header
+
+
+class System(Base):
+    def __init__(self):
+        self.request = self.get_header()
+
+    def get_data(self, raw):
+        """
+        Extracts key=value pairs with float/integer from ntp response packet data.
+        """
+        data = dict()
+        for key, value in self.regex.findall(raw):
+            data[key] = float(value) * PRECISION
+        return data
+
+
+class Peer(Base):
+    def __init__(self, idx, name):
+        self.id = idx
+        self.real_name = name
+        self.name = name.replace('.', '_')
+        self.request = self.get_header(self.id)
+
+    def get_data(self, raw):
+        """
+        Extracts key=value pairs with float/integer from ntp response packet data.
+        """
+        data = dict()
+        for key, value in self.regex.findall(raw):
+            dimension = '_'.join([self.name, key])
+            data[dimension] = float(value) * PRECISION
+        return data
+
+
+class Service(SocketService):
+    def __init__(self, configuration=None, name=None):
+        SocketService.__init__(self, configuration=configuration, name=name)
+        self.order = list(ORDER)
+        self.definitions = dict(CHARTS)
+
+        self.port = 'ntp'
+        self.dgram_socket = True
+        self.system = System()
+        self.peers = dict()
+        self.request = str()
+        self.retries = 0
+        self.show_peers = self.configuration.get('show_peers', False)
+
+    def check(self):
+        """
+        Checks if we can get valid systemvars.
+        If not, returns None to disable module.
+        """
+        self._parse_config()
+
+        if self.show_peers:
+            self.definitions.update(PEER_CHARTS)
+
+            peer_filter = self.configuration.get('peer_filter', r'127\..*')
+            try:
+                self.peer_filter = re.compile(r'^((0\.0\.0\.0)|({0}))$'.format(peer_filter))
+            except re.error as error:
+                self.error('Compile pattern error (peer_filter) : {0}'.format(error))
+                return None
+
+            try:
+                self.peer_rescan = int(self.configuration.get('peer_rescan', peer_rescan))
+                if self.peer_rescan <= 0:
+                    raise ValueError('int > 0 expected: {0}'.format(self.peer_rescan))
+            except ValueError as error:
+                self.error('Value error (peer_rescan) : {0}'.format(error))
+                return None
```

Comment:
please move to init()
```python
self.configuration.get('peer_rescan', 60)
```

That is all. All checks are not needed.

Reply:
@l2isbad OK. I did this check, because I saw when you put `peer_rescan: 0` in the config it will result in an unhandled exception:

```
python.d ERROR: ntpd: ntpd: update() unhandled exception: integer division or modulo by zero
```

Reply:
@l2isbad and leave `peer_filter` check in `check()` or also move it to `init()`?

Reply:
`update() unhandled exception` is ok :smile: 

by the description of the option it is clear that it's must be `> 1` (you can add comment about it in conf file)

> peer_filter check in check()

well, do as you want. If you feel it should be in `check()` leave it in `check()` :+1: